src/details/ArborX_DetailsHDBSCAN.hpp

```diff
@@ -0,0 +1,401 @@
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_DETAILS_HDBSCAN_HPP
+#define ARBORX_DETAILS_HDBSCAN_HPP
+
+#include <ArborX_DetailsMutualReachabilityDistance.hpp>
+#include <ArborX_DetailsTreeNodeLabeling.hpp>
+#include <ArborX_LinearBVH.hpp>
+
+namespace ArborX
+{
+
+namespace Details
+{
+
+// Boruvka's algorithm is guaranteed to converge only if all edges have unique
+// weights. As the distances between pairs of points may not be unique, point
+// indices are used to resolve edges uniquely.
+KOKKOS_INLINE_FUNCTION
+bool compareEdgesLess(int i1, int j1, float d1, int i2, int j2, float d2)
+{
+  using KokkosExt::max;
+  using KokkosExt::min;
+
+  if (d1 < d2)
+    return true;
+  if (d1 > d2)
+    return false;
+
+  auto vmin1 = min(i1, j1);
+  auto vmin2 = min(i2, j2);
+  if (vmin1 < vmin2)
+    return true;
+  if (vmin1 > vmin2)
+    return false;
+
+  auto vmax1 = max(i1, j1);
+  auto vmax2 = max(i2, j2);
+  if (vmax1 < vmax2)
+    return true;
+  return false;
+}
+
+struct EdgeTriple
+{
+  int i = -1;
+  int j = -1;
+  float d = KokkosExt::ArithmeticTraits::infinity<float>::value;
+};
+
+KOKKOS_FUNCTION inline bool operator<(EdgeTriple const &edge1,
+                                      EdgeTriple const &edge2)
+{
+  if (compareEdgesLess(edge1.i, edge1.j, edge1.d, edge2.i, edge2.j, edge2.d))
+    return true;
+  return false;
+}
+
+template <typename ExecutionSpace, typename Predicates, typename BVH,
+          typename BVHParents, typename Labels, typename CoreDistances,
+          typename ComponentOutEdges>
+void determineComponentEdges(ExecutionSpace const &exec_space,
+                             Predicates const &predicates, BVH const &bvh,
+                             BVHParents const &bvh_parents,
+                             Labels const &labels,
+                             CoreDistances const &core_distances,
+                             ComponentOutEdges components_out_edges)
+{
+  Kokkos::Profiling::pushRegion("ArborX::HDBSCAN::determine_component_edges");
+
+  using MemorySpace = typename Labels::memory_space;
+  using Access = AccessTraits<Predicates, PredicatesTag>;
+
+  auto const n = Access::size(predicates);
+
+  ARBORX_ASSERT(labels.size() == n);
+
+  auto NoInit = [](std::string const &label) {
+    return Kokkos::view_alloc(Kokkos::WithoutInitializing, label);
+  };
+
+  Kokkos::View<int *, MemorySpace> bvh_nodes_labels(
+      NoInit("ArborX::HDBSCAN::labels"), 2 * n - 1);
```

Comment:
Why do you have both an array with the labels of the leaf nodes in the original order and here another array for both internal and leaf nodes?

Reply:
The `labels` array is the final output with cluster labels. Here, it is an array only used during Boruvka iterations, with labels corresponding to found Boruvka components. They are totally different. As this PR is only the first phase, we don't actually construct `labels` in any form (that will be done in the final phase when computing flat clustering out of a dendrogram based on stability metrics).